// @appium/doctor/build/lib/prompt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clear = exports.fixIt = exports.configure = void 0;
const utils_1 = require("./utils");
/**
 * @type {string}
 */
let persistentResponse;
const fixItQuestion = {
    type: 'list',
    name: 'confirmation',
    message: 'Fix it:',
    choices: ['yes', 'no', 'always', 'never'],
    /**
     *
     * @param {string} val
     * @returns {string}
     */
    filter(val) {
        return val.toLowerCase();
    },
};
function configure(opts) {
    if (opts.yes) {
        persistentResponse = 'yes';
    }
    if (opts.no) {
        persistentResponse = 'no';
    }
}
exports.configure = configure;
function clear() {
    persistentResponse = undefined;
}
exports.clear = clear;
async function fixIt() {
    if (persistentResponse) {
        return persistentResponse;
    }
    let resp = await utils_1.inquirer.prompt(fixItQuestion);
    persistentResponse = resp.confirmation === 'always' ? 'yes' : persistentResponse;
    persistentResponse = resp.confirmation === 'never' ? 'no' : persistentResponse;
    return persistentResponse || resp.confirmation;
}
exports.fixIt = fixIt;
